### PR TITLE
Return pooled blocks when connection is idle

### LINKED ISSUE #699: Return unused block from SocketInput/Output when connection is idle
`SocketInput` and `SocketOutput` both have `_head` and `_tail` fields referencing `MemoryPoolBlock`s. In both cases, the bytes between `_head` and `_tail` haven't been consumed.

When all the bytes are fully consumed `_head == _tail` and the `Start` and `End` properties of the block are also equal. Still the fully consumed block is never returned to the `MemoryPool` until the connection is closed. This causes each idle connection to waste 8KB of memory even when the write buffers are completely flushed and the read allocation callback hasn't been called.

### PR DESCRIPTION
**Pull request by halter73**
Fixes #699 

To test the reduction in memory usage, I ran the following console app against the Kestrel sample with an argument of 10000. I simply looked at task manager's private working set memory for the process after all the sockets connect. Not the most scientific, but it showed a consistent memory reduction from ~220KB to ~130KB.

    using System;
    using System.Net;
    using System.Net.Sockets;
    using System.Text;

    namespace ConsoleApp1
    {
        public class Program
        {
            private static Socket[] _sockets;

            public static void Main(string[] args)
            {
                var numSocket = int.Parse(args[0]);
                _sockets = new Socket[numSocket];

                var requestString = "GET / HTTP/1.1\r\n\r\n";
                byte[] requestBytes = Encoding.ASCII.GetBytes(requestString);
                byte[] responseBytes = new byte[1000];

                for (int i = 0; i < numSocket; i++)
                {
                    _sockets[i] = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                    _sockets[i].Connect(new IPEndPoint(IPAddress.Loopback, 5000));
                    _sockets[i].Send(requestBytes);
                    _sockets[i].Receive(responseBytes);

                    if (i % 100 == 99)
                    {
                        Console.WriteLine($"{i + 1} sockets connected");
                    }
                }

                Console.WriteLine($"{numSocket} sockets connected to http://localhost:5000");
                Console.ReadLine();
            }
        }
    }

## Before

image

## After

image

### CODE REVIEW

**Comment by benaadams** [comment]
Does this release the block while waiting in keepalive? (I think you are saying it does?)

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs [hunk 1]
```diff
@@ -172,18 +178,31 @@ public MemoryPoolIterator ConsumingStart()
...
                         // Must call Subtract() after _head has been advanced, to avoid producer starting too early and growing
                         // buffer beyond max length.
                         _bufferSizeControl?.Subtract(lengthConsumed);
                     }
 
-                    if (!examined.IsDefault &&
-                        examined.IsEnd &&
-                        ReadingInput)
+                    // If _head is null, everything has been consumed and examined.
+                    var examidedAll = (!examined.IsDefault && examined.IsEnd) || _head == null;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: examided

**Comment by halter73** [comment]
@benaadams yes. This releases blocks whenever the input is fully consumed or the output is fully written. So before idle, keep-alive connections would previously keep 2 4KB blocks leased until the connection was close. Now the 8KB worth of blocks will be returned to the pool once the last request/response is complete.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/SockAddr.cs [hunk 2]
```diff
@@ -7,14 +7,18 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Networking
 {
     public struct SockAddr
     {
+#pragma warning disable CS0414
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/SockAddr.cs]: No need for this - see #1123

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Networking/SockAddr.cs]: Saw that. More elegant for sure.

**Comment by halter73** [comment]
@cesarbs @pakrym ping

**Comment by halter73** [comment]
@pakrym Same ballpark with plaintext RPS:

    benchmarks\src\BenchmarksDriver [dev]> dotnet run -- -n plaintext -s http://adx-3u-perfsvr:5001 -c http://adx-3u-svr02:5002 -o KestrelHttpServer@dev
    Project Repository (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ClientJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ServerJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project BenchmarksDriver (.NETCoreApp,Version=v1.0) was previously compiled. Skipping compilation.
    [03:45:14.098] Starting scenario Plaintext on benchmark server...
    [03:45:36.641] Warmup
    [03:45:36.653] Starting scenario Plaintext on benchmark client...
    [03:45:47.736] Scenario Plaintext completed on benchmark client
    [03:45:47.736] RPS: 1123011.21
    [03:45:47.736] Stopping scenario Plaintext on benchmark client...
    [03:45:47.738] Benchmark
    [03:45:47.738] Starting scenario Plaintext on benchmark client...
    [03:45:58.781] Scenario Plaintext completed on benchmark client
    [03:45:58.782] RPS: 1165965.48
    [03:45:58.782] Stopping scenario Plaintext on benchmark client...
    [03:45:58.784] Stopping scenario Plaintext on benchmark server...
    benchmarks\src\BenchmarksDriver [dev]> dotnet run -- -n plaintext -s http://adx-3u-perfsvr:5001 -c http://adx-3u-svr02:5002 -o KestrelHttpServer@halter73/699
    Project Repository (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ClientJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ServerJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project BenchmarksDriver (.NETCoreApp,Version=v1.0) was previously compiled. Skipping compilation.
    [03:46:06.012] Starting scenario Plaintext on benchmark server...
    [03:46:28.628] Warmup
    [03:46:28.641] Starting scenario Plaintext on benchmark client...
    [03:46:39.728] Scenario Plaintext completed on benchmark client
    [03:46:39.729] RPS: 1173892.26
    [03:46:39.730] Stopping scenario Plaintext on benchmark client...
    [03:46:39.734] Benchmark
    [03:46:39.734] Starting scenario Plaintext on benchmark client...
    [03:46:50.788] Scenario Plaintext completed on benchmark client
    [03:46:50.788] RPS: 1202135.6
    [03:46:50.788] Stopping scenario Plaintext on benchmark client...
    [03:46:50.789] Stopping scenario Plaintext on benchmark server...

**Comment by pakrym** [comment]
:shipit:

**Comment by halter73** [comment]
I messed up the merge ordering. The 2 commits in this PR are https://github.com/aspnet/KestrelHttpServer/commit/0edf36bd2159a69bbdcab5a806be904b4649d2fd and https://github.com/aspnet/KestrelHttpServer/commit/7b2f7b94ab56dd78411d6d29e3a5d9598e042c0f.

**Pull request closed without merging**